extension/goplugin/js_integration_test.go

```diff
@@ -0,0 +1,128 @@
+// implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package goplugin_test
+
+import (
+	"io"
+	"os"
+	"time"
+
+	"github.com/cloudwan/gohan/extension"
+	"github.com/cloudwan/gohan/extension/goext"
+	"github.com/cloudwan/gohan/extension/goplugin"
+	"github.com/cloudwan/gohan/extension/otto"
+	l "github.com/cloudwan/gohan/log"
+	"github.com/cloudwan/gohan/schema"
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+)
+
+var _ = Describe("JS integration", func() {
+	var (
+		env              *goplugin.Environment
+		schemaManager    *schema.Manager
+		extensionManager *extension.Manager
+		logWriter        io.Writer = os.Stderr
+	)
+
+	const (
+		schemaPath = "test_data/test_schema.yaml"
+	)
+
+	BeforeSuite(func() {
+		l.SetUpBasicLogging(logWriter, l.DefaultFormat)
+	})
+
+	initGlobalObjects := func() {
+		schemaManager = schema.GetManager()
+		Expect(schemaManager.LoadSchemaFromFile(schemaPath)).To(Succeed())
+
+		extensionManager = extension.GetManager()
+	}
+
+	initGoExtension := func() {
+		env = goplugin.NewEnvironment("test", nil, nil)
+
+		err := env.Load("test_data/ext_good/ext_good.so")
+		Expect(err).To(BeNil())
+		Expect(env.Start()).To(Succeed())
+	}
+
+	initOttoExtension := func() {
+		ottoEnv := otto.NewEnvironment("otto", nil, nil, nil)
+		ottoCode, err := schema.NewExtension(map[string]interface{}{
+			"id": "test_otto_extension",
+			"code": `gohan_register_handler("js_throw_custom_exception", function(context) {
+			      	 	gohan_log_debug("JS throwing handler invoked");
+						context.js_invoked = true;
+      					throw new CustomException("test exception", 456);
+					});
+			`,
+			"path": ".*",
+		})
+		Expect(err).ToNot(HaveOccurred())
+		ottoCode.URL = "good_extension.js"
+
+		extensions := []*schema.Extension{ottoCode}
+		timeLimit := time.Duration(1) * time.Minute
+		timeLimits := []*schema.PathEventTimeLimit{}
+		Expect(ottoEnv.LoadExtensionsForPath(extensions, timeLimit, timeLimits, "test")).To(Succeed())
+
+		extensionManager.RegisterEnvironment("test", ottoEnv)
+	}
+
+	BeforeEach(func() {
+		initGlobalObjects()
+		initGoExtension()
+		initOttoExtension()
+
+		testSchema := env.Schemas().Find("test")
+		Expect(testSchema).NotTo(BeNil())
+	})
+
+	AfterEach(func() {
+		env.Stop()
+	})
+
+	Context("Error propagation", func() {
+		var (
+			ctx goext.Context
+			err error
+		)
+
+		BeforeEach(func() {
+			ctx = map[string]interface{}{
+				"schema_id": "test",
+			}
+
+			err = env.Core().TriggerEvent("js_throw_custom_exception", ctx)
+		})
+
+		It("should propagate errors", func() {
+			Expect(err).To(HaveOccurred())
+			Expect(ctx).To(HaveKeyWithValue("js_invoked", true))
+		})
+
+		It("should clean context", func() {
+			Expect(ctx).NotTo(HaveKey("exception"))
+			Expect(ctx).NotTo(HaveKey("exception_message"))
+		})
+
+		It("should propagate error code", func() {
+			Expect(err.(goext.Error).Code).To(Equal(goext.ErrorCode(456)))
```

Comment:
the error sould be *goext.Error since it is a pointer
also, check if message has also been passed correctly

Reply:
goext.NewError returns an goext.Error, not *goext.Error. I'm aware this will change in #543, but that PR is not yet merged.
Depending on which PR is merged sooner, this will require a change, but that change will be guarded by the test, so we won't forget about it.

Reply:
ok